python/src/utils.cpp

```diff
@@ -0,0 +1,63 @@
+
+#include "mlx/utils.h"
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <optional>
+
+namespace py = pybind11;
+using namespace py::literals;
+using namespace mlx::core;
+
+// Slightly different from the original, with python context on init we are not
+// in the context yet. Only create the inner context on enter then de   lete on
+// exit.
+class PyStreamContextManager {
+ public:
+  PyStreamContextManager(StreamOrDevice s) : _s(s), _inner(nullptr) {}
+
+  void enter() {
+    _inner = new StreamContextManager(_s);
+  }
+
+  void exit() {
+    if (_inner != nullptr) {
+      delete _inner;
+      _inner = nullptr;
+    }
+  }
+
+ private:
+  StreamOrDevice _s;
+  StreamContextManager* _inner;
+};
+
+void init_utils(py::module_& m) {
+  py::class_<PyStreamContextManager>(m, "StreamContextManager", R"pbdoc(
+        A context manager for setting the current device and stream.
+    
+        This class is a Python wrapper around the C++ class
+        `mlx::core::StreamContextManager`. It is used to set the current device
```

Comment:
I don't think there is a need for that you can directly say that it can change the default device and stream for the duration of the context as you have it. No need to mention that it wraps the equivalent to the C++ object.